src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,342 @@
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+#include "UsfDecoderPlugin.hxx"
+#include "../DecoderAPI.hxx"
+#include "../DecoderInternal.hxx"
+#include "tag/TagHandler.hxx"
+#include "fs/Path.hxx"
+#include "util/Domain.hxx"
+#include "Log.hxx"
+
+#include <usf.h>
+#include <psflib.h>
+
+static constexpr Domain usf_domain("usf");
+static constexpr unsigned USF_BUFFER_FRAMES = 2048;
+static constexpr unsigned USF_CHANNELS = 2;
+static constexpr unsigned USF_BUFFER_SAMPLES = USF_BUFFER_FRAMES*USF_CHANNELS;
+
+static void * stdio_fopen( const char * path )
+{
+    return fopen( path, "rb" );
+}
+
+static size_t stdio_fread( void *p, size_t size, size_t count, void *f )
+{
+    return fread( p, size, count, (FILE*) f );
+}
+
+static int stdio_fseek( void * f, int64_t offset, int whence )
+{
+    return fseek( (FILE*) f, offset, whence );
+}
+
+static int stdio_fclose( void * f )
+{
+    return fclose( (FILE*) f );
+}
+
+static long stdio_ftell( void * f )
+{
+    return ftell( (FILE*) f );
+}
+
+psf_file_callbacks stdio_callbacks =
+{
+    "\\/:",
+    stdio_fopen,
+    stdio_fread,
+    stdio_fseek,
+    stdio_fclose,
+    stdio_ftell
+};
+
+struct usf_loader_state
+{
+    uint32_t enable_compare;
+    uint32_t enable_fifo_full;
+    char * lib;
+    void * emu;
+};
+
+static int
+usf_loader(void * context, const uint8_t * exe, size_t exe_size,
+               const uint8_t * reserved, size_t reserved_size)
+{
+    struct usf_loader_state * state = ( struct usf_loader_state * ) context;
+
+    if ( exe && exe_size > 0 ) return -1;
+
+    return usf_upload_section( state->emu, reserved, reserved_size );
+}
+
+static int
+usf_info(void * context, const char * name, const char * value)
+{
+    struct usf_loader_state * state = ( struct usf_loader_state * ) context;
+
+    if ( strcasecmp( name, "_enablecompare" ) == 0 && strlen( value ) )
+        state->enable_compare = 1;
+    else if ( strcasecmp( name, "_enablefifofull" ) == 0 && strlen( value ) )
+        state->enable_fifo_full = 1;
+
+    return 0;
+}
+
+struct usf_tags {
+    std::string title;
+    std::string artist;
+    std::string composer;
+    std::string game;
+    std::string year;
+    std::string copyright;
+    std::string usfby;
+    std::string genre;
+    std::string length;
+    std::string fade;
+    std::string track;
+};
+
+static int
+usf_save_tag(const char * field, const char * name, const char *value, std::string& tag)
+{
+    if (strcasecmp(name, field) == 0) {
+        tag = value;
+        return 1;
+    }
+    return 0;
+}
+
+static int
+usf_tags_target(void * context, const char * name, const char * value)
+{
+    struct usf_tags * tags = ( struct usf_tags * ) context;
+    if (!value) return 0;
+
+    if (usf_save_tag("title", name, value, tags->title));
```

Comment:
Empty block (only `;`) after `if` is bad style, and is an invitation for hard-to-find bugs.